src/ConvNetSharp.Core/Training/TrainerBase.cs

```diff
@@ -15,35 +15,35 @@ protected TrainerBase(INet<T> net)
 
         public double BackwardTimeMs { get; private set; }
 
-        public T CostLoss { get; private set; }
-
         public double ForwardTimeMs { get; private set; }
 
-        public virtual T Loss => this.CostLoss;
+        public double UpdateWeightsTimeMs { get; private set; }
+
+        public virtual T Loss { get; private set; }
 
         public int BatchSize { get; set; } = 1;
 
+        public void ResetTimes()
+        {
+            ForwardTimeMs = 0;
+            BackwardTimeMs = 0;
+            UpdateWeightsTimeMs = 0;
+        }
+
         protected virtual void Backward(Volume<T> y)
         {
             var chrono = Stopwatch.StartNew();
 
-            if (typeof(T) == typeof(double))
-            {
-                this.CostLoss = (T) (object) ((double) (object) this.Net.Backward(y) / y.Shape.GetDimension(3));
-            }
-            else if (typeof(T) == typeof(float))
-            {
-                this.CostLoss = (T) (object) ((float) (object) this.Net.Backward(y) / y.Shape.GetDimension(3));
-            }
-
-            this.BackwardTimeMs = chrono.Elapsed.TotalMilliseconds / y.Shape.GetDimension(3);
+            var dimension = Ops<T>.Cast(y.Shape.GetDimension(3));
+            this.Loss = Ops<T>.Divide(this.Net.Backward(y), dimension);
+            this.BackwardTimeMs += chrono.Elapsed.TotalMilliseconds;
```

Comment:
I was experimenting with using average time across all samples, because using time per sample was being to erratic. I reverted back to time per sample, and removed the ResetTime method since it's no longer needed.